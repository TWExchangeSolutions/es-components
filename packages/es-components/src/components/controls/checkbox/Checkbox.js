import React from 'react';
import PropTypes from 'prop-types';
import styled, { ThemeProvider, withTheme } from 'styled-components';
import { noop } from 'lodash';
import viaTheme from 'es-components-via-theme';

import Label from '../Label';

/* eslint-disable no-confusing-arrow */
const CheckboxLabel = Label.extend`
  font-size: ${props => props.theme.sizes.baseFontSize}px;
  font-weight: normal;
  text-transform: none;
  display: flex;

  > .checkbox-fill {
    background-color: ${props =>
    props.isChecked ? props.theme.colors.info : props.theme.colors.white};
    border-color: ${props =>
    props.isChecked ? props.theme.colors.info : props.theme.colors.grayDark};

    &:after {
      border-color: ${props => props.theme.colors.white};
    }
  }

  &:hover > .checkbox-fill:after {
    border-color: ${props =>
    props.isChecked
      ? props.theme.colors.white
      : props.theme.colors.grayLight};
  }

  &[disabled] > .checkbox-fill {
    background-color: ${props =>
    props.isChecked ? props.theme.colors.gray : props.theme.colors.white};
    border-color: ${props => props.theme.colors.gray};
    cursor: not-allowed;
    outline: 0;

    &:after {
      border-color: ${props => props.theme.colors.white};
    }
  }
`;

const CheckboxInput = styled.input`
  clip: rect(0, 0, 0, 0);
  position: absolute;

  &:focus ~ .checkbox-fill {
    box-shadow: 0 0 3px 3px ${props => props.theme.colors.inputFocus};
    &:after {
      border-color: ${props =>
    props.checked
      ? props.theme.colors.white
      : props.theme.colors.grayLight};
    }
  }
`;
/* eslint-enable */

const CheckboxWrapper = styled.span`
  background: ${props => props.theme.colors.white};
  border: 3px solid ${props => props.theme.colors.grayDarker};
  border-radius: 4px;
  box-sizing: border-box;
  cursor: pointer;
  height: 25px;
  transition: all 0.25s linear;
  width: 25px;

  &:after {
    background: transparent;
    border: 3px solid ${props => props.theme.colors.white};
    border-right: none;
    border-top: none;
    box-sizing: border-box;
    content: '';
    height: 9px;
    display: block;
    transform: rotate(-45deg);
    transition: border 0.25s linear;
    width: 15px;
    margin: 3px 0 0 2px;
  }
`;

const CheckboxText = styled.span`
  line-height: 1.4;
  margin-left: 5px;
`;

function Checkbox({
  labelText,
  value,
  isChecked = false,
  isDisabled = false,
  onClick = noop,
  onChange = noop,
  theme
}) {
  /* eslint-disable jsx-a11y/use-onblur-not-onchange */
  return (
    <ThemeProvider theme={theme}>
      <CheckboxLabel disabled={isDisabled} isChecked={isChecked}>
        <CheckboxInput
          type="checkbox"
          disabled={isDisabled}
          value={value}
          checked={isChecked}
          onClick={onClick}
          onChange={onChange}
          focusBorderColor={theme.colors.inputFocus}
        />
        <CheckboxWrapper className="checkbox-fill" />
        <CheckboxText>{labelText}</CheckboxText>
      </CheckboxLabel>
    </ThemeProvider>
  );
  /* eslint-enable */
}

Checkbox.propTypes = {
  labelText: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  isChecked: PropTypes.bool,
  /* Function to execute when a checkbox is clicked */
  onClick: PropTypes.func,
  /* Function to execute when a checkbox checked state is changed */
  onChange: PropTypes.func,
  isDisabled: PropTypes.bool,
  /**
   * Theme object used by the ThemeProvider,
   * automatically passed by any parent component using a ThemeProvider
   */
  theme: PropTypes.object
};

Checkbox.defaultProps = {
  theme: viaTheme
};

export default withTheme(Checkbox);
