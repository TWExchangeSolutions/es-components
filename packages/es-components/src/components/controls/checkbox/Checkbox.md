```
<div>
  <Checkbox labelText="I am unchecked by default" />
  <Checkbox labelText="I am checked by default" isChecked />
  <Checkbox labelText="I am disabled" isDisabled />
  <Checkbox labelText="I am checked and disabled" isChecked isDisabled />
</div>
```

Passing a function will execute with the current selection status of the Checkbox.

```
class CheckboxExample extends React.Component {
  constructor() {
    this.state = { 
      isAppleChecked: false,
      isBananaChecked: false,
      isStrawberryChecked: true
    }

    this.toggleCheckbox = this.toggleCheckbox.bind(this);
  }

  toggleCheckbox(key) {
    return event => {
      this.setState({ [key]: event.target.checked });
    }
  }

  render() {
    return (
      <div>
        <Checkbox 
          labelText="Apple"
          value="Apple"
          isChecked={this.state.isAppleChecked}
          onClick={this.toggleCheckbox('isAppleChecked')}
        />
        <Checkbox 
          labelText="Banana"
          value="Banana"
          isChecked={this.state.isBananaChecked}
          onClick={this.toggleCheckbox('isBananaChecked')}
        />
        <Checkbox 
          labelText="Strawberry"
          value="Strawberry"
          isChecked={this.state.isStrawberryChecked}
          onClick={this.toggleCheckbox('isStrawberryChecked')}
          isDisabled
        />
      </div>
    );
  }
}

<CheckboxExample />
```