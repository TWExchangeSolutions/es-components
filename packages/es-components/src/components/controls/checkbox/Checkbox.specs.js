/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import Checkbox from './Checkbox';

let instance;

function simulateLabelClick() {
  instance.find('input').simulate('click', { target: { checked: true } });
}

beforeEach(() => {
  instance = mount(<Checkbox labelText="test" />);
});

it('the onClick prop gets executed with the value of the checkbox when clicked', () => {
  const onClick = jest.fn();
  instance.setProps({ onClick });

  simulateLabelClick();

  expect(onClick).toHaveBeenCalled();
});

it('Checkbox renders as expected', () => {
  const tree = renderer.create(
    <Checkbox labelText="Render test" isDisabled isChecked />
  );
  expect(tree).toMatchSnapshot();
});
