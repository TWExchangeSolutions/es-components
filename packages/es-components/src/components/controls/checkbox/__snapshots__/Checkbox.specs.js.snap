// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders as expected 1`] = `
<label
  className="sc-bwzfXH hizqnz"
  disabled={true}
>
  <input
    checked={true}
    className="sc-htpNat hESkBQ"
    disabled={true}
    onChange={[Function]}
    onClick={[Function]}
    type="checkbox"
    value={undefined}
  />
  <span
    className="checkbox-fill sc-bxivhb fwlpgh"
  />
  <span
    className="sc-ifAKCX kVENxO"
  >
    Render test
  </span>
</label>
`;
